AppleEnrollment/BulkAssignProfile.ps1

```diff
@@ -0,0 +1,509 @@
+
+        }
+
+        $id = [Guid]::NewGuid();
+        if([string]::IsNullOrWhiteSpace($ProfileId) -or ![Guid]::TryParse($ProfileId, [ref]$id)){
+
+            write-host "Invalid ProfileId specified, please specify valid ProfileId to assign to devices..." -f Red
+
+        }
+        elseif ($Devices -eq $null -or $Devices.Count -eq 0){
+            
+            write-host "No devices specified, please specify a list of devices to assign..." -f Red
+        }
+        else {
+
+            $Resource = [String]::Format("deviceManagement/enrollmentProfiles('{0}')/updateDeviceProfileAssignment", $ProfileId)
+
+            $JSON = [String]@"
+            `"deviceIds`": [
+                {0}
+            ]
+"@
+
+            [String]$d = ""
+            Foreach($device in $Devices)
+            {
+                if([String]::IsNullOrWhiteSpace($d))
+                {
+                    $d = "`"" + $device + "`""
+                }
+                else
+                {
+                    $d = $d + ",`"" + $device + "`""
+                }
+            }
+
+            $JSON = "{" + [String]::Format($JSON, $d) + "}"
+
+            Test-JSON -JSON $JSON
+
+            $uri = "https://graph.microsoft.com/$graphApiVersion/$($Resource)"
+            Invoke-RestMethod -Uri $uri -Headers $authToken -Method Post -Body $JSON -ContentType "application/json"
+            
+            Write-Host "Devices assigned!" -f Green
+        }
+
+    }
+
+    catch {
+
+    Write-Host
+    $ex = $_.Exception
+    $errorResponse = $ex.Response.GetResponseStream()
+    $reader = New-Object System.IO.StreamReader($errorResponse)
+    $reader.BaseStream.Position = 0
+    $reader.DiscardBufferedData()
+    $responseBody = $reader.ReadToEnd();
+    Write-Host "Response content:`n$responseBody" -f Red
+    Write-Error "Request to $Uri failed with HTTP Status $($ex.Response.StatusCode) $($ex.Response.StatusDescription)"
+    write-host
+    break
+
+    }
+
+}
+
+####################################################
+
+Function Get-UnAssignedDevices(){
+
+<#
+.SYNOPSIS
+This function is used to get all un-assigned bulk devices using the Graph API REST interface
+.DESCRIPTION
+The function connects to the Graph API Interface and gets all un-assigned bulk devices
+.EXAMPLE
+Get-UnAssignedDevices
+Gets all un-assigned bulk devices
+.NOTES
+NAME: Get-UnAssignedDevices
+#>
+
+[cmdletbinding()]
+
+param
+(
+)
+
+$graphApiVersion = "Beta"
+$ResourceSegment = "deviceManagement/importedAppleDeviceIdentities?`$filter=DiscoverySource eq 'deviceEnrollmentProgram'"
+
+    try {
+
+        [System.String]$skipToken = ''
+        [System.String[]]$unAssignedDevices = @()
+        [System.Uri]$uri = "https://graph.microsoft.com/$graphApiVersion/$($ResourceSegment)"
+
+        DO
+        {
+            $response = Invoke-RestMethod -Uri $uri -Headers $authToken -Method Get -ContentType "application/json"
+            $skipToken = Get-SkipToken([System.string]$response)
```

Comment:
I would recommend just using the odata.nextLink property rather than looking through the response for a string match, as this code could potentially break in future if we change paging here's some example code

```powershell
$usersNextLink = $usersResponse."@odata.nextLink"
while ($usersNextLink -ne $null) {
    $usersResponse = (Invoke-RestMethod -Uri $usersNextLink -Headers $authHeader -Method Get -Verbose)
    $usersNextLink = $usersResponse."@odata.nextLink"
    $users += $usersResponse.value
}
```